### PR TITLE
use python gpg

### PR DESCRIPTION
**Pull request by pazz**
This is a rebase of #906 against current master.
please make sure I didn't introduce anything silly!

### CODE REVIEW

File: alot/crypto.py [hunk 1]
```diff
@@ -194,15 +155,14 @@ def verify_detached(message, signature):
...
     :raises: :class:`~alot.errors.GPGProblem` if the verification fails
     '''
-    message_data = StringIO(message)
-    signature_data = StringIO(signature)
-    ctx = gpgme.Context()
+    ctx = gpg.Context()
     try:
-        return ctx.verify(signature_data, message_data, None)
-    except gpgme.GpgmeError as e:
+        (data, verify_results) = ctx.verify(gpg.Data(string=message), gpg.Data(string=signature), message_data)
```
Comment (lucc) [inline alot/crypto.py]: what is `message_data` doing here?

File: alot/crypto.py [hunk 2]
```diff
@@ -49,7 +19,7 @@ def RFC3156_micalg_from_algo(hash_algo):
     :rtype: str
     """
     # hash_algo will be something like SHA256, but we need pgp-sha256.
-    hash_algo = _hash_algo_name(hash_algo)
+    hash_algo = gpg._gpgme.gpgme_hash_algo_name(hash_algo)
```
Comment (lucc) [inline alot/crypto.py]: Is it possible to di this with some public method?

Comment (zenhack) [inline alot/crypto.py]: Any reason for not having a wrapper? (can we request one?)

The leading underscore by convention denotes private in python, so if this is going to go in as-is I'd at least want a comment explaining the situation (perhaps linking to @teythoon's comment). Ideally upstream would expose this and then we could just call the wrapper.

File: alot/crypto.py [hunk 3]
```diff
@@ -154,18 +124,14 @@ def detached_signature_for(plaintext_str, key=None):
 
     :param plaintext_str: text to sign
     :param key: gpgme_key_t object representing the key to use
-    :rtype: tuple of gpgme.NewSignature array and str
+    :rtype: tuple of gpg.results.NewSignature array and str
     """
-    ctx = gpgme.Context()
+    ctx = gpg.Context()
     ctx.armor = True
```
Comment (teythoon) [inline alot/crypto.py]: Or: gpg.Context(armor=True).  Also gpg.Context is a context manager.

**Comment by pazz** [comment]
Just to mention: I did not test this so far! (didn't find the time to figure out how to install this on debian testing)

File: alot/crypto.py [hunk 4]
```diff
@@ -154,18 +124,14 @@ def detached_signature_for(plaintext_str, key=None):
...
     if key is not None:
         ctx.signers = [key]
-    plaintext_data = StringIO(plaintext_str)
-    signature_data = StringIO()
-    sigs = ctx.sign(plaintext_data, signature_data, gpgme.SIG_MODE_DETACH)
-    signature_data.seek(0, os.SEEK_SET)
-    signature = signature_data.read()
-    return sigs, signature
+    (sigblob, sign_result) = ctx.sign(gpg.Data(string=plaintext_str), mode=gpg.constants.SIG_MODE_DETACH)
```
Comment (teythoon) [inline alot/crypto.py]: Actually, we wrap flos with a fileno() method and everything supporting the buffer protocol magically (in Python2 this includes strings, in Python3 bytestrings), no need for an explicit wrapping here.

... = ctx.sign(plaintext_str, ...) should suffice here.

Comment (teythoon) [inline alot/crypto.py]: ... and everywhere else of course.

File: alot/crypto.py (hunk 2 above)
Comment (teythoon) [inline alot/crypto.py]: Sure.  The reason is that no one has written a wrapper.  We are a free software project, and welcome contributions.

**Comment by teythoon** [comment]
Patrick Totzke <notifications@github.com> writes:

> Just to mention: I did not test this so far! (didn't find the time to
> figure out how to install this on debian testing)

You can try 'pip install gpg'.

**Comment by pazz** [comment]
I'm *such* a fan of pip :P I'll figure it out (currently i have a broken outdated local pip install)

**Comment by teythoon** [comment]
Patrick Totzke <notifications@github.com> writes:

> I'm *such* a fan of pip :P

Hah, me too :p However, I was convinced that one has to be on pypi to be
taken seriously, so we are.

**Comment by pazz** [comment]
Quoting Justus Winter (2016-12-12 17:50:39)
>    Patrick Totzke <notifications@github.com> writes:
> 
>    > I'm *such* a fan of pip :P
> 
>    Hah, me too :p However, I was convinced that one has to be on pypi to be
>    taken seriously, so we are.

OK, thanks for this :)

Sorry but I have to ask: any Idea which package I should install on deb testing to
satisfy this dependency?

```
$pip install gpg                                                              1 ↵  
Collecting gpg
  Using cached gpg-1.8.0.tar.gz
    Complete output from command python setup.py egg_info:
    Could not find gpgme-config.  Please install the libgpgme development package.
    
    ----------------------------------------
Command "python setup.py egg_info" failed with error code 1 in /tmp/pip-build-i_5eD2/gpg/
```


I have installed `python-gpgme` and there is no other package that contains "gpgme" in its name..

**Comment by lucc** [comment]
have a look at the old travis PR I put up to find the potential name of the libgpgme dev package I think I installed that on ubuntu on travis. It should(?) be the same name on debian.

**Comment by teythoon** [comment]
Patrick Totzke <notifications@github.com> writes:

> Quoting Justus Winter (2016-12-12 17:50:39)
>>    Patrick Totzke <notifications@github.com> writes:
>> 
>>    > I'm *such* a fan of pip :P
>> 
>>    Hah, me too :p However, I was convinced that one has to be on pypi to be
>>    taken seriously, so we are.
>
> OK, thanks for this :)
>
> Sorry but I have to ask: any Idea which package I should install on deb testing to
> satisfy this dependency?

libgpgme11-dev

**Comment by pazz** [comment]
Quoting Justus Winter (2016-12-13 09:40:11)
>    > Sorry but I have to ask: any Idea which package I should install on deb
>    testing to
>    > satisfy this dependency?
> 
>    libgpgme11-dev

That's it, thanks. I only had to install "swig" as well and all the
build errors were gone :)
I can test this branch now.

**Comment by pazz** [comment]
I think there is a bug in the quantifiedcode integration.
The merge is blocked here due to "waiting for status...",
but the log on their site reads 
```
...
[INFO / 2016-12-14 08:29:20] Analyzing pull request 915...
[INFO / 2016-12-14 08:29:20] Skipping this pull request as it is still up-to-date.
```

**Comment by pazz** [comment]
tested a bit just now. There is a bug triggered when I want to open a thread containing a signed mail:

```
--- <exception caught here> ---
  File "/usr/lib/python2.7/dist-packages/twisted/internet/defer.py", line 149, in maybeDeferred
    result = f(*args, **kw)
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/commands/search.py", line 38, in apply
    sb = buffers.ThreadBuffer(ui, self.thread)
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/buffers.py", line 334, in __init__
    self.rebuild()
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/buffers.py", line 362, in rebuild
    self._tree = ThreadTree(self.thread)
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/widgets/thread.py", line 357, in __init__
    accumulate(msg)
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/widgets/thread.py", line 336, in accumulate
    self._message[mid] = MessageTree(msg, odd)
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/widgets/thread.py", line 169, in __init__
    self._maintree = SimpleTree(self._assemble_structure())
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/widgets/thread.py", line 196, in _assemble_structure
    mainstruct.append((self._get_headers(), None))
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/widgets/thread.py", line 259, in _get_headers
    headers)
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/widgets/thread.py", line 273, in construct_header_pile
    mail = self._message.get_email()
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/db/message.py", line 74, in get_email
    self._email = message_from_file(f_mail)
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/db/utils.py", line 140, in message_from_file
    m.get_payload(1).get_payload())
  File "/home/pazz/.local/lib/python2.7/site-packages/alot-0.4.dev0-py2.7.egg/alot/crypto.py", line 163, in verify_detached
    (data, verify_results) = ctx.verify(gpg.Data(string=message), gpg.Data(string=signature), message_data)
exceptions.NameError: global name 'message_data' is not defined
```

**Comment by dkg** [comment]
I've just rebased my own use-python-gpg branch, but i don't appear to have push access to put the branch into the pazz/alot repo itself.  Please feel free to force-push to the pazz/alot repo, or to grant me permission to do a non-fast-forward push to the dkg-use-python-gpg-906 branch on this repo.

**Comment by pazz** [comment]
@dkg: could you just open a new PR for this? I think it'd be easiest.
If you are interested, I can give you maintainance rights to this repository though.

**Comment by dkg** [comment]
On Thu 2017-01-12 16:02:15 -0500, Patrick Totzke wrote:
> @dkg: could you just open a new PR for this? I think it'd be easiest.
> If you are interested, I can give you maintainance rights to this
> repository though.

sure, i can open a new PR, but it seems complicated to have to follow
chains of PRs to understand the evolution of a patchset.  I don't want
maintenance rights to the repo if you don't want them, but it might be
nice to be able to push to this one branch at least :)

     --dkg

**Comment by pazz** [comment]
Quoting dkg (2017-01-13 15:17:19)
>    On Thu 2017-01-12 16:02:15 -0500, Patrick Totzke wrote:
>    > @dkg: could you just open a new PR for this? I think it'd be easiest.
>    > If you are interested, I can give you maintainance rights to this
>    > repository though.
> 
>    sure, i can open a new PR, but it seems complicated to have to follow
>    chains of PRs to understand the evolution of a patchset. I don't want
>    maintenance rights to the repo if you don't want them, but it might be
>    nice to be able to push to this one branch at least :)

Oh I wasn't going to ditch the responsibility onto you, just offering
you write access :).
I will try to figure out how to give you push rights to an individual
branch, but I suspect thats not so easy. stay tuned :)

**Comment by pazz** [comment]
@dkg: It seems that I cannot give you push access just to this one branch without giving you push rights to the whole repo (i.e., set you up as collaborator). It's your choice, you seem to have a pending invitation to be a collaborator.

**Comment by pazz** [comment]
@dkg: could you upload your updated branch somewhere?

File: alot/crypto.py [hunk 5]
```diff
@@ -3,39 +3,9 @@
 # For further details see the COPYING file
 import os
 from cStringIO import StringIO
```
Comment (meskio) [inline alot/crypto.py]: Unused imports `os` and `StringIO`.

File: alot/crypto.py (hunk 1 above)
Comment (meskio) [inline alot/crypto.py]: I'm assuming it should be `message`.

File: alot/crypto.py [hunk 6]
```diff
@@ -212,45 +172,25 @@ def decrypt_verify(encrypted):
...
     :raises: :class:`~alot.errors.GPGProblem` if the decryption fails
     '''
-    encrypted_data = StringIO(encrypted)
-    plaintext_data = StringIO()
-    ctx = gpgme.Context()
+    ctx = gpg.Context()
     try:
-        sigs = ctx.decrypt_verify(encrypted_data, plaintext_data)
-    except gpgme.GpgmeError as e:
+        (plaintext, result, verify_result) = ctx.decrypt(gpg.Data(string=encrypted_data), verify=True)
```
Comment (meskio) [inline alot/crypto.py]: I think should be `s/encrypted_data/encrypted/`

File: alot/crypto.py [hunk 7]
```diff
@@ -212,45 +172,25 @@ def decrypt_verify(encrypted):
...
+    except gpg.errors.GPGMEError as e:
         raise GPGProblem(e.message, code=e.code)
```
Comment (meskio) [inline alot/crypto.py]: s/e.code/e.getcode()/

File: alot/crypto.py (hunk 2 above)
Comment (teythoon) [inline alot/crypto.py]: Actually, there already is a wrapper for that.  gpg.core.hash_algo_name.  Sorry for the noise.

Comment (zenhack) [inline alot/crypto.py]: Cool, we should use it then.

**Comment by jljusten** [comment]
@dkg: I think if you force push your repo's use-python-gpg branch, then it should update the #906 pull request.

**Comment by pazz** [comment]
Apologies for the chaos! I've accidentally removed this branch to push a re-based version 
(https://github.com/pazz/alot/tree/dkg-use-python-gpg-906) and cannot reopen this PR.
@meskio : is this commit https://github.com/dkg/alot/pull/1 relevant and should go in? 
Do we have some version of this that you guys agreed on so far?

**Comment by pazz** [comment]
FWIW: my current rebase fails the travis tests: https://travis-ci.org/pazz/alot/jobs/235940262
The docs job fails die to an unmet dependency to gpg, which should be an easy fix.
The other I'm not sure about..

**Comment by meskio** [comment]
@pazz yes, I believe is relevant. gpg.verify can throw a `BadSignature` exception and we should act properly on it.

I'm not sure this is ready to merge, last time I tested it somethings where failing (I don't recall what).

**Comment by jljusten** [comment]
@pazz: Are you still looking into this? Or, do you need @dkg to update his branch?

It sounds like debian may remove pygpgme at some point.
https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=846314

**Comment by pazz** [comment]
I'm not actively working on this. So everybody's welcome to push this along. ultimately we do want to replace pygpgme by gpg i suppose

**Comment by dcbaker** [comment]
I went ahead and opened an issue for this to help us track it better.

**Comment by pazz** [comment]
Thanks! Do you want to close this one then?
P

**Comment by dcbaker** [comment]
This one is closed already, thus I opened the issue.

**Comment by jljusten** [comment]
link to the issue @dcbaker opened: https://github.com/pazz/alot/issues/1069

**Pull request closed without merging**